File: src/migration_rules/format_string_rule/before_script.py
#!/usr/bin/env python3
"""
Format String Migration - BEFORE Script (Spark 3.2 and earlier)
This script demonstrates the deprecated 0$ indexing in format_string and printf functions.

Usage:
    python before_script.py

This script can be run independently to test the deprecated behavior.
"""

from pyspark.sql import SparkSession
from pyspark.sql.functions import expr

def main():
    # Create SparkSession
    spark = SparkSession.builder \
        .appName('FormatStringBefore') \
        .master('local[*]') \
        .getOrCreate()
    
    print("=== Format String Migration - BEFORE (Deprecated 0$ indexing) ===")
    
    try:
        # Create sample DataFrame
        data = [
            ('John', 5, 'Premium'),
            ('Alice', 100, 'Gold'),
            ('Bob', 25, 'Silver'),
            ('Carol', 75, 'Premium')
        ]
        df = spark.createDataFrame(data, ['name', 'message_count', 'tier'])
        
        print("\nOriginal DataFrame:")
        df.show()
        
        # Using format_string with 0$ indexing (deprecated in Spark 3.3+)
        print("\n1. Using format_string with 0$ indexing:")
        result1 = df.select(
            'name',
            'message_count',
            expr("format_string('Hello %0$s, you have %0$d messages', name, message_count)").alias('greeting')
        )
        result1.show(truncate=False)
        
        # Using printf with 0$ indexing (deprecated in Spark 3.3+)
        print("\n2. Using printf with 0$ indexing:")
        result2 = df.select(
            'name',
            'tier',
            expr("printf('User %0$s has %0$s tier status', name, tier)").alias('user_info')
        )
        result2.show(truncate=False)
        
        # Complex format string with multiple 0$ references
        print("\n3. Complex format string with multiple 0$ references:")
        result3 = df.select(
            'name',
            'message_count',
            'tier',
            expr("format_string('Welcome %0$s! You are a %0$s member with %0$d unread messages', name, tier, message_count)").alias('welcome_msg')
        )
        result3.show(truncate=False)
        
        # Nested format string operations
        print("\n4. Nested format string operations with 0$ indexing:")
        result4 = df.select(
            'name',
            'message_count',
            expr("format_string('Status: %0$s', format_string('User %0$s has %0$d items', name, message_count))").alias('nested_format')
        )
        result4.show(truncate=False)
        
        print("\n⚠️  WARNING: This code uses deprecated 0$ indexing which is not supported in Spark 3.3+")
        print("   It may cause parsing errors or unexpected behavior in newer versions.")
        print("   Expected errors: ArgumentIndexOutOfBoundsException or similar parsing errors")
        
    except Exception as e:
        print(f"❌ Error executing format string operations: {e}")
        print("   This error is expected in Spark 3.3+ due to deprecated 0$ indexing")
        print("   The error demonstrates why migration to 1$-based indexing is necessary")
    
    finally:
        spark.stop()
        print("\nSpark session stopped.")

if __name__ == "__main__":
    main()

File: src/migration_rules/format_string_rule_scala/before_script.scala
// Format String Argument Index Migration - BEFORE Script (Spark 3.2 and earlier)
// This script demonstrates the old behavior where format_string() and printf() supported 0$ indexing.
//
// Language: Scala
// Usage: spark-shell -i before_script.scala
//
// This script can be run independently to test the old behavior.

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._

object FormatStringMigrationBefore {
  def main(args: Array[String]): Unit = {
    // Create SparkSession
    val spark = SparkSession.builder()
      .appName("FormatStringMigrationBefore")
      .master("local[*]")
      .config("spark.sql.warehouse.dir", "spark-warehouse")
      .getOrCreate()

    import spark.implicits._

    println("=== Format String Argument Index Migration - BEFORE (Spark 3.2 behavior) ===")

    try {
      // Create sample data
      println("\n📊 Creating sample data...")
      val data = Seq(
        ("Alice", 25, 50000.0),
        ("Bob", 30, 60000.0),
        ("Charlie", 35, 70000.0)
      ).toDF("name", "age", "salary")

      data.show()

      // Demonstrate the problematic 0$ indexing in format_string
      println("\n⚠️  ATTEMPTING: Using 0$ indexing in format_string (Spark 3.2 behavior)...")
      println("   SQL: SELECT format_string('Employee: %0$s, Age: %1$d', name, age)")

      try {
        // This would work in Spark 3.2 and earlier
        val result1 = data.select(
          col("name"),
          col("age"),
          expr("format_string('Employee: %0$s, Age: %1$d', name, age)").alias("formatted_0_index")
        )

        println("✅ SUCCESS: 0$ indexing worked in Spark 3.2")
        result1.show(truncate = false)

      } catch {
        case e: Exception =>
          println(s"❌ Error with 0$ indexing: ${e.getMessage}")
          println("   This error indicates you're running on Spark 3.3+ where 0$ indexing is not supported")
      }

      // Demonstrate the problematic 0$ indexing in printf
      println("\n⚠️  ATTEMPTING: Using 0$ indexing in printf...")
      println("   SQL: SELECT printf('Salary: %0$.2f for %1$s', salary, name)")

      try {
        val result2 = data.select(
          col("name"),
          col("salary"),
          expr("printf('Salary: %0$.2f for %1$s', salary, name)").alias("printf_0_index")
        )

        println("✅ SUCCESS: 0$ indexing in printf worked in Spark 3.2")
        result2.show(truncate = false)

      } catch {
        case e: Exception =>
          println(s"❌ Error with 0$ indexing in printf: ${e.getMessage}")
          println("   This demonstrates the breaking change in Spark 3.3")
      }

      // Show more complex example with multiple 0$ references
      println("\n⚠️  ATTEMPTING: Complex format string with multiple 0$ references...")
      println("   SQL: SELECT format_string('%0$s is %1$d years old and earns %2$.2f. %0$s works hard!', name, age, salary)")

      try {
        val result3 = data.select(
          col("*"),
          expr("format_string('%0$s is %1$d years old and earns %2$.2f. %0$s works hard!', name, age, salary)")
            .alias("complex_0_index")
        )

        println("✅ SUCCESS: Complex 0$ indexing worked")
        result3.show(truncate = false)

        println("\n⚠️  LEGACY BEHAVIOR ISSUES:")
        println("   • 0$ indexing was non-standard and confusing")
        println("   • Inconsistent with most printf implementations")
        println("   • Could lead to off-by-one errors in argument positioning")
        println("   • Made format strings harder to read and maintain")

      } catch {
        case e: Exception =>
          println(s"❌ Error: ${e.getMessage}")
          println("   Complex 0$ indexing patterns are no longer supported")
      }

      // Demonstrate DataFrame API usage
      println("\n📊 Using DataFrame API with format_string...")
      try {
        val dfResult = data.withColumn(
          "description",
          expr("format_string('Name: %0$s, Details: Age %1$d, Salary $%2$.2f', name, age, salary)")
        )

        println("✅ DataFrame API with 0$ indexing:")
        dfResult.select("name", "description").show(truncate = false)

      } catch {
        case e: Exception =>
          println(s"❌ DataFrame API error: ${e.getMessage}")
      }

    } catch {
      case e: Exception =>
        println(s"❌ Error during demonstration: ${e.getMessage}")
        println("   This demonstrates compatibility issues with format string indexing")
    } finally {
      spark.stop()
      println("\nSpark session stopped.")

      println("\n🎯 SUMMARY - Spark 3.2 and Earlier Behavior:")
      println("   • format_string() and printf() supported 0$ indexing")
      println("   • First argument could be referenced as %0$")
      println("   • This was non-standard compared to other printf implementations")
      println("   • Could cause confusion and maintenance issues")
    }
  }
}

// If running in spark-shell, call the main method
FormatStringMigrationBefore.main(Array())